type-directed compilation

flupe opened the pull request:
A big refactoring as a first step to properly handle module parameters.

The most important change is the following: the compilation of terms if now completely type-directed. That is, we no longer rely on `ArgInfo` of `Elims` to know whether they have to be preserved, but rather their `Dom Type` only.

What this means, concretely:
- Explicit usable erasable arguments (Say, `(a : Set)`) are dropped.
- Implicit usable non-erasable arguments (Say, `{x : Nat}`) are made explicit.  
  (This means we now officially support tactic arguments that are made explicit on the Haskell side.)

As a treat, we now properly erase types of sort `Prop`.

Closes #217.

---

TODO list before opening the PR:

- [x] Add back special class functions.
- [x] Fix compilation of co-patterns.
- [x] Fix compilation of patterns on projection-like functions.
- [x] Add back compilation of inline functions.
- [x] Ensure unfolding is back.
- [x] Drop module parameters again.

flupe commented:
All the tests pass again, finally.

Before reviewing this PR, there's still one caveat to fix: currently every test has the flag `--no-projection-like` set, because projection-like are annoying to manipulate.

The following things are likely to break once I remove the flag:
- [ ] Compilation of functions, as projection-like only store patterns that come after the record argument.
- [ ] Compilation of `Def`s, because some defs will be compiled in `compileProj`.
- [ ] `compileSpined`?

jespercockx commented on src/Agda2Hs/Compile/Type.hs:
```diff
@@ -210,16 +209,30 @@ compileInlineType f args = do
+
+data DomOutput = DOInstance | DODropped | DOKept
+
+compiledDom :: Dom Type -> C DomOutput
```
Why is this a separate function from `compileDom`? Seems like it's a potential source of subtle bugs if these two don't match.

flupe replied:
The reason was `compileDom` atually compiles all the types to Haskell, so if this was the one being used in the typed-compilation we would be constantly compiling types to Haskell, even though we only care about whether those things are kept or dropped.
If we had dependent types it would be easier to express at the type-level that `compileDom` is now a refinement of `compiledDom`, but I'll just implement the former with the latter.

jespercockx commented on src/Agda2Hs/Compile.hs:
```diff
@@ -24,81 +25,67 @@ import Agda2Hs.Compile.Types
-  { currModule = tlm
-  , minRecordName = Nothing
-  , locals = []
+  { currModule        = tlm
```
It would be a bit nicer to bundle all cosmetic changes such as this one in a separate PR, since it will make it easier to locate any bugs that might be introduced.

jespercockx commented on src/Agda2Hs/Compile/Utils.hs:
```diff
@@ -46,6 +47,36 @@ import Agda2Hs.Compile.Types
+    loop h es =
+      case es of
+        []            -> pure Nothing
+        Proj o f : es -> undefined -- Just $ fromMaybe (Def f (Apply (defaultArg v) : es')) $ loop h (Proj o f : res) es'
```
This looks like it should still be fixed.

jespercockx commented on src/Agda2Hs/Compile/Utils.hs:
```diff
@@ -46,6 +47,36 @@ import Agda2Hs.Compile.Types
+  case tm of
+    -- Just (h, es) -> fromMaybe v $ loop h [] es
+    Var i es     -> fromMaybe (tm, ty) <$> loop (Var i) es
+    Def f es     -> undefined 
```
This looks like it should still be fixed.

jespercockx commented on src/Agda2Hs/Compile/Utils.hs:
```diff
@@ -287,3 +330,56 @@ checkSingleElement name fs s = unless (length fs == 1) $ genericDocError =<< do
 checkingVars :: C a -> C a
 checkingVars = local $ \e -> e { checkVar = True }
+
+checkFixityLevel :: QName -> FixityLevel -> C (Maybe Int)
```
Could this stuff related to fixities perhaps be moved to a separate PR?

jespercockx commented on src/Agda2Hs/Render.hs:
```diff
@@ -75,11 +76,15 @@ moduleFileName opts name = do
   outDir <- compileDir
   return $ fromMaybe outDir (optOutDir opts) </> moduleNameToFileName name "hs"
 
+-- NOTE: what does this have to do with rendering?
```
Should probably be moved to the top-level Compile module

jespercockx commented on test/UnboxPragma.agda:
```diff
@@ -22,7 +22,7 @@ sort1 xs = xs [ looksfine ]
 sort2 xs .el = xs
 sort2 xs .pf = looksfine
-
+ 
```
Please fix whitespace

jespercockx commented on src/Agda2Hs/Compile/Function.hs:
```diff
@@ -163,103 +165,122 @@ compileFun' withSig def@(Defn {..}) = do
+  -- NOTE: should be fine for unboxed records
+  unlessM (asks copatternsEnabled `or2M` (isJust <$> isUnboxProjection pn)) $
+    genericDocError =<< "not supported in Haskell: copatterns"
+  Just (unEl -> Pi a b) <- getDefType pn ty -- ????
```
Shouldn't you use `mustBePi` here to account for possible reduction?

jespercockx commented on src/Agda2Hs/Compile/Term.hs:
```diff
@@ -20,311 +21,446 @@ import Agda.Syntax.Internal
+
+  case compileSpined ty v of
+    Just cont -> cont
+    Nothing   -> case unSpine1 v of
```
Is `unSpine1` here still necessary? All cases with a spine are already handled by `compileSpined`.

jespercockx reviewed (commented):
Great work! This is definitely a major improvement to the compilation process of `agda2hs`. Going through the changes also made me realize that we're really not dealing with a small project here anymore. I have a couple of remarks and I think the projection-like functions should still be fixed. But overall I think this should be merged sooner rather than later.

flupe commented:
Rebased on #277.

omelkonian commented on src/Agda2Hs/Compile/Term.hs:
```diff
@@ -20,313 +21,448 @@ import Agda.Syntax.Internal
+expTypeSig _ _ = genericError "`the` must be fully applied"
+
+
+-- should really be named compileVar, TODO: rename compileVar
```
I suggest `s/compileVar/compileVarName` and then `s/compileV/compileVar`.

flupe replied:
`compileVar` as renamed to `compileDBVar` in #277 . Renamed `compileV` to `compileVar`, lmk if that's enough.

omelkonian reviewed (commented):
Great work Lucas!

One thing that slightly worries me is that I would expect this refactoring to reduce code size, but it in fact increases it substantially (c.f. [+800/-500](https://github.com/flupe/agda2hs/compare/cosmetic...flupe:agda2hs:modules)). Factoring out new tests and utilities we did not need before, it seems the main culprit is `Compile.Term`; do you think the final polished version might be able to reduce LOC down a bit?

PS: Ping me when the projection-like optimization has been properly handled.

jakobn-ai commented:
Implicit arguments with instance functions are still a compiler error, is this expected? i.e.

```agda
test : {x : Nat} → Nat
test {x} = x
{-# COMPILE AGDA2HS test #-}
```

will work but

```agda
instance
  test : {x : Nat} → Nat
  test {x} = x
{-# COMPILE AGDA2HS test #-}
```

will give

```
An internal error has occurred. Please report this as a bug.
Location of the error: __IMPOSSIBLE__, called at src/Agda2Hs/Compile/ClassInstance.hs:103:20 in main:Agda2Hs.Compile.ClassInstance
```

flupe commented:
Hmm, to be fair the compilation of class instances is the part I'm the least happy with currently (the current implementation and what I did to make it typed).

Though I fail to see when a *compiled* instance with hidden relevant arguments turning relevant would ever make sense. I will add a check that for instances specifically, no relevant arguments are ever hidden as I think this would do the trick. Thanks for the example!

If you ever encounter an example where this would be useful behavior, hit me up!

flupe commented:
@omelkonian: I fixed the compilation of projection-like functions and addressed all of your comments!

You will see that one test still requires the `--no-projection-like` flag as compiling projection-like (or projections) directly applied to constructors in a typed setting kinda goes against bidirectional tc. Quoting @jespercockx:

> a projection-like function applied to a constructor violates bidirectionality so you cannot (easily) do a type-aware traversal of it

If you remove the flag, the code will still compile to correct Haskell, but the projection-like function `unbind` will get reduced. Not incorrect, but just not what the user wrote.

If someone is really motivated, perhaps using [projView](https://hackage.haskell.org/package/Agda-2.6.4.1/docs/Agda-TypeChecking-ProjectionLike.html#v:projView) could allow you to preserve the projection-like function application, but I think you'd still have to trigger inference at some point, which I don't want to do.

> One thing that slightly worries me is that I would expect this refactoring to reduce code size, but it in fact increases it substantially (c.f. [+800/-500](https://github.com/flupe/agda2hs/compare/cosmetic...flupe:agda2hs:modules)). Factoring out new tests and utilities we did not need before, it seems the main culprit is Compile.Term; do you think the final polished version might be able to reduce LOC down a bit?

Hm, going into this implementation I had the same assumption, but quickly realized some more work was unavoidable (e.g Defs and projections cannot really be treated uniformly anymore).
All in all, I don't think too much code is added, there is also a lot of whitespace introduced because I like my code to breathe.

Hopefully things should be more easily maintainable too :)

----

>  @jakobn-ai: came across this experimenting with adding runtime checks for e.g. erased preconditions. A deciding version of a precondition would ideally be written as instance so it can be found in instance search: [...]

But the instance itself shouldn't be compiled to agda2hs right? Or using some custom logic (that you'd have to implement yourself).

I decided against adding the check I mentioned above, since the class compilation routine already doesn't enforce that, say, we don't pattern-match on parameters of instances, and they are silently dropped til we reach a (class) projection ([See here](https://github.com/agda/agda2hs/blob/4e3b9a73c87c03140f95d817fee29618333d1dcb/src/Agda2Hs/Compile/ClassInstance.hs#L138)).

So something should be done alright, just not in this PR!

omelkonian commented:
> If you remove the flag, the code will still compile to correct Haskell, but the projection-like function `unbind` will get reduced. Not incorrect, but just not what the user wrote.

Could you also add these unit tests, i.e. the ones without the flag where the generated code is further away from what the user wrote?

Then, a subsequent PR that improves on that (e.g. using `projView`) could improve those tests.

flupe commented:
Done! (See `test/ProjLike.agda`)
